File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9E.scala
package TextbookTrack.Chapter09

object BA9E {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}

  sealed trait NodeColor
  final case object Blue extends NodeColor
  final case object Red extends NodeColor
  final case object Purple extends NodeColor
  final case object Gray extends NodeColor

  final case class Edge(nodeTo: Int, startIx: Int, length: Int)

  class SuffixTree(text1: String, text2: String) {
    val text: String = text1 + text2

    private val text1Size: Int = text1.length
    private val size: Int = text.length
    private val nodeIdGenerator: Iterator[Int] = Iterator.from(0)

    val root: Int = nodeIdGenerator.next()
    val adjacencyList: Map[Int, List[Edge]] = buildSuffixTree()

    lazy val nrNodes: Int =
      adjacencyList.valuesIterator.flatten.foldLeft(0){ case (n, Edge(nodeTo, _, _)) => math.max(n, nodeTo) } + 1

    def doesEdgeBelongToFirstText(edge: Edge): Boolean = edge.startIx < text1Size

    private def buildSuffixTree(): Map[Int, List[Edge]] = {
      val adjacencyList: MutableMap[Int, List[Edge]] = MutableMap()
      (0 until size).foreach(addSuffixToTree(adjacencyList, _))
      adjacencyList.toMap
    }

    private def addSuffixToTree(adjacencyList: MutableMap[Int, List[Edge]], suffixStart: Int): Unit = {
      val (node, optionalEdge, edgeStartIx, nonMatchingIx): (Int, Option[Edge], Int, Int) =
        findLastMatchingNode(adjacencyList, suffixStart)
      val newLeaf: Int = nodeIdGenerator.next()
      optionalEdge match {
        case None =>
          adjacencyList
            .update(node, Edge(newLeaf, edgeStartIx, size - edgeStartIx) :: adjacencyList.getOrElse(node, Nil))
        case Some(edge @ Edge(nodeTo, startIx, length)) =>
          val newNode: Int = nodeIdGenerator.next()
          val splitLength1: Int = nonMatchingIx - edgeStartIx
          val splitLength2: Int = length - splitLength1
          adjacencyList(node) = adjacencyList(node).filterNot(_ == edge)
          adjacencyList.update(node, Edge(newNode, startIx, splitLength1) :: adjacencyList.getOrElse(node, Nil))
          adjacencyList(newNode) =
            List(Edge(nodeTo, startIx + splitLength1, splitLength2), Edge(newLeaf, nonMatchingIx, size - nonMatchingIx))
      }
    }

    private def findLastMatchingNode(adjacencyList: MutableMap[Int, List[Edge]],
                                     suffixStart: Int): (Int, Option[Edge], Int, Int) = {
      def findNextEdgeInPath(node: Int, letter: Char): Option[Edge] =
        adjacencyList.getOrElse(node, Nil).find{ case Edge(_, startIx, _) => text(startIx) == letter }

      @tailrec
      def loop(currentNode: Int, ix: Int): (Int, Option[Edge], Int, Int) =
        findNextEdgeInPath(currentNode, text(ix)) match {
          case None => (currentNode, None, ix, ix)
          case Some(edge @ Edge(nodeTo, _, length)) =>
            findFirstNonMatchingIndex(ix, edge) match {
              case Some(nonMatchingIx) => (currentNode, Some(edge), ix, nonMatchingIx)
              case None => loop(nodeTo, ix + length)
            }
        }

      loop(root, suffixStart)
    }

    private def findFirstNonMatchingIndex(ix: Int, edge: Edge): Option[Int] = {
      val Edge(_, startIx, length) = edge
      val substring: String = text.slice(startIx, startIx + length)
      substring
        .lazyZip(text.substring(ix))
        .lazyZip(ix until size)
        .find{ case (a, b, _) => a != b }
        .map{ case (_, _, jy) => jy }
    }
  }

  private def performNodeColoring(suffixTree: SuffixTree): Vector[NodeColor] = {
    val nodeColors: Array[NodeColor] = initializeNodeColors(suffixTree)

    @tailrec
    def loop(ripeNodes: List[Int]): Unit = {
      if (ripeNodes.nonEmpty) {
        ripeNodes.foreach{
          node =>
            val children: Set[Edge] = suffixTree.adjacencyList.getOrElse(node, Nil).toSet
            val colors: Set[NodeColor] = children.map{ case Edge(nodeTo, _, _) => nodeColors(nodeTo) }
            nodeColors(node) = if (colors.size == 1) colors.iterator.next() else Purple
        }
        loop(findRipeNodes(suffixTree, nodeColors))
      }
    }

    loop(findRipeNodes(suffixTree, nodeColors))
    nodeColors.toVector
  }

  private def initializeNodeColors(suffixTree: SuffixTree): Array[NodeColor] = {
    val nodeColors: Array[NodeColor] = Array.fill(suffixTree.nrNodes)(Gray)

    @tailrec
    def loop(currentNodes: List[(Int, NodeColor)]): Unit = {
      if (currentNodes.nonEmpty) {
        val (leaves, innerNodes): (List[((Int, NodeColor), List[Edge])], List[((Int, NodeColor), List[Edge])]) =
          currentNodes
            .map{ case pair @ (node, _) => (pair, suffixTree.adjacencyList.getOrElse(node, Nil)) }
            .partition{ case (_, edges) => edges.isEmpty }

        leaves.foreach{ case ((node, nodeColor), _) => nodeColors(node) = nodeColor }
        val nextNodes: List[(Int, NodeColor)] =
          innerNodes.flatMap{ case (_, edges) => collectEdgeEndpoints(suffixTree, edges) }
        loop(nextNodes)
      }
    }

    loop(List((suffixTree.root, Gray)))
    nodeColors
  }

  private def collectEdgeEndpoints(suffixTree: SuffixTree, edges: List[Edge]): List[(Int, NodeColor)] =
    edges.map{ case edge @ Edge(nodeTo, _, _) =>
      if (suffixTree.doesEdgeBelongToFirstText(edge)) (nodeTo, Blue) else (nodeTo, Red)
    }

  private def findRipeNodes(suffixTree: SuffixTree, nodeColors: Array[NodeColor]): List[Int] =
    nodeColors
      .iterator
      .zipWithIndex
      .filter{ case (color, _) => color == Gray }
      .filter{ case (_, node) =>
        val children = suffixTree.adjacencyList.getOrElse(node, Nil)
        children.forall{ case Edge(nodeTo, _, _) => nodeColors(nodeTo) != Gray }
      }
      .map{ case (_, node) => node }
      .toList

  private def findLongestSubstringSpelledByPurpleNodes(suffixTree: SuffixTree,
                                                       nodeColors: Vector[NodeColor]): String = {
      @tailrec
      def loop(longestSharedSubstring: String, currentNodes: List[(Int, String)]): String = {
        if (currentNodes.isEmpty) longestSharedSubstring
        else {
          val nextNodes: List[(Int, String)] =
            currentNodes.flatMap{
              case (node, spelledString) =>
                getPurpleNeighbourNodes(suffixTree, nodeColors, node, spelledString)
            }
          loop(updateLongestSharedSubstring(longestSharedSubstring, nextNodes), nextNodes)
        }
      }

      val emptyString: String = ""
      loop(emptyString, List((suffixTree.root, emptyString)))
    }

  private def getPurpleNeighbourNodes(suffixTree: SuffixTree,
                                      nodeColors: Vector[NodeColor],
                                      node: Int,
                                      spelledString: String): List[(Int, String)] = {
    suffixTree
      .adjacencyList
      .getOrElse(node, Nil)
      .withFilter{ case Edge(nodeTo, _, _) => nodeColors(nodeTo) == Purple }
      .map{ case Edge(nodeTo, startIx, length) =>
        (nodeTo, spelledString + suffixTree.text.slice(startIx, startIx + length))
      }
  }

  private def updateLongestSharedSubstring(substring: String, nodes: List[(Int, String)]): String =
    nodes.foldLeft(substring){ case (acc, (_, string)) => if (string.length > acc.length) string else acc }

  def findLongestSharedSubstring(text1: String, text2: String): String = {
    val suffixTree = new SuffixTree(text1 + '#', text2 + '$')
    val nodeColors: Vector[NodeColor] = performNodeColoring(suffixTree)
    findLongestSubstringSpelledByPurpleNodes(suffixTree, nodeColors)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text1: String = reader.next()
    val text2: String = reader.next()
    val result: String = findLongestSharedSubstring(text1, text2)
    println(result)
  }
}

File: Python/textbook_track/chapter09/ba9e.py
# Find the Longest Substring Shared by Two Strings
import functools as ft
import itertools as it
import sys

from collections import defaultdict, namedtuple
from enum import Enum, auto

Edge = namedtuple('Edge', ['node_to', 'substring_start', 'substring_length'])


class NodeColor(Enum):
    BLUE = auto()
    RED = auto()
    PURPLE = auto()
    GRAY = auto()


class SuffixTree:
    def __init__(self, text1, text2):
        self._text = text1 + text2
        self._text1_size = len(text1)
        self._size = len(self._text)
        self._node_id_generator = it.count()
        self._root = next(self._node_id_generator)
        self._adjacency_list = self._build_suffix_tree()

    @property
    def text(self):
        return self._text

    @property
    def root(self):
        return self._root

    @property
    def adjacency_list(self):
        return self._adjacency_list

    @ft.cached_property
    def nr_nodes(self):
        n = 0
        for edges in self._adjacency_list.values():
            for edge in edges:
                if edge.node_to > n:
                    n = edge.node_to
        return n + 1

    def does_edge_belong_to_first_text(self, edge):
        return edge.substring_start < self._text1_size

    def _build_suffix_tree(self):
        adjacency_list = defaultdict(list)
        for ix in range(self._size):
            self._add_suffix_to_tree(adjacency_list, ix)
        return dict(adjacency_list)

    def _add_suffix_to_tree(self, adjacency_list, suffix_start):
        node, edge, edge_start_ix, non_matching_ix = self._find_last_matching_node(adjacency_list, suffix_start)
        new_leaf = next(self._node_id_generator)
        if edge is None:
            adjacency_list[node].append(Edge(new_leaf, edge_start_ix, self._size - edge_start_ix))
        else:
            new_node = next(self._node_id_generator)
            split_length_1 = non_matching_ix - edge_start_ix
            split_length_2 = edge.substring_length - split_length_1
            adjacency_list[node].remove(edge)
            adjacency_list[node].append(Edge(new_node, edge.substring_start, split_length_1))
            adjacency_list[new_node].append(Edge(edge.node_to, edge.substring_start+split_length_1, split_length_2))
            new_edge = Edge(new_leaf, non_matching_ix, self._size - non_matching_ix)
            adjacency_list[new_node].append(new_edge)

    def _find_last_matching_node(self, adjacency_list, suffix_start):
        ix = suffix_start
        current_node = self._root
        while (edge := self._find_next_edge_in_path(adjacency_list, current_node, self._text[ix])) is not None:
            non_matching_ix = self._find_first_non_matching_index(ix, edge)
            if non_matching_ix is not None:
                return current_node, edge, ix, non_matching_ix

            current_node = edge.node_to
            ix += edge.substring_length

        return current_node, None, ix, ix

    def _find_next_edge_in_path(self, adjacency_list, node, letter):
        edges = adjacency_list.get(node, [])
        for edge in edges:
            if self._text[edge.substring_start] == letter:
                return edge
        return None

    def _find_first_non_matching_index(self, ix, edge):
        substring = self._text[edge.substring_start:edge.substring_start+edge.substring_length]
        for jy, (a, b) in enumerate(zip(self._text[ix:], substring), start=ix):
            if a != b:
                return jy
        return None


def perform_node_coloring(suffix_tree):
    node_colors = initialize_node_colors(suffix_tree)
    while (ripe_nodes := find_ripe_nodes(suffix_tree, node_colors)):
        for node in ripe_nodes:
            children = suffix_tree.adjacency_list.get(node, [])
            colors = {node_colors[edge.node_to] for edge in children}
            if len(colors) == 1:
                node_colors[node] = next(iter(colors))
            else:
                node_colors[node] = NodeColor.PURPLE
    return tuple(node_colors)


def initialize_node_colors(suffix_tree):
    node_colors = [NodeColor.GRAY] * suffix_tree.nr_nodes
    current_nodes = [(suffix_tree.root, NodeColor.GRAY)]
    while current_nodes:
        next_nodes = []
        for node, node_color in current_nodes:
            edges = suffix_tree.adjacency_list.get(node, [])
            if edges:
                neighbours = collect_edge_endpoints(suffix_tree, edges)
                next_nodes.extend(neighbours)
            else:
                node_colors[node] = node_color
        current_nodes = next_nodes
    return node_colors


def collect_edge_endpoints(suffix_tree, edges):
    neighbours = []
    for edge in edges:
        if suffix_tree.does_edge_belong_to_first_text(edge):
            node_color = NodeColor.BLUE
        else:
            node_color = NodeColor.RED
        neighbours.append((edge.node_to, node_color))
    return neighbours


def find_ripe_nodes(suffix_tree, node_colors):
    ripe_nodes = []
    for node, color in enumerate(node_colors):
        if color == NodeColor.GRAY:
            children = suffix_tree.adjacency_list.get(node, [])
            if all(node_colors[edge.node_to] != NodeColor.GRAY for edge in children):
                ripe_nodes.append(node)
    return ripe_nodes


def find_longest_substring_spelled_by_purple_nodes(suffix_tree, node_colors):
    longest_shared_substring = ''
    current_nodes = [(suffix_tree.root, longest_shared_substring)]
    while current_nodes:
        next_nodes = []
        for node, spelled_string in current_nodes:
            next_nodes.extend(get_purple_neighbour_nodes(suffix_tree, node_colors, node, spelled_string))

        longest_shared_substring = update_longest_shared_substring(longest_shared_substring, next_nodes)
        current_nodes = next_nodes

    return longest_shared_substring


def get_purple_neighbour_nodes(suffix_tree, node_colors, node, spelled_string):
    purple_children = []
    edges = suffix_tree.adjacency_list.get(node, [])
    for neighbour, substring_start, substring_length in edges:
        if node_colors[neighbour] == NodeColor.PURPLE:
            edge_label = suffix_tree.text[substring_start:substring_start+substring_length]
            purple_children.append((neighbour, spelled_string + edge_label))
    return purple_children


def update_longest_shared_substring(longest_shared_substring, nodes):
    length = len(longest_shared_substring)
    for _, string in nodes:
        if len(string) > length:
            longest_shared_substring = string
            length = len(longest_shared_substring)

    return longest_shared_substring


def find_longest_shared_substring(text1, text2):
    suffix_tree = SuffixTree(text1 + '#', text2 + '$')
    node_colors = perform_node_coloring(suffix_tree)
    return find_longest_substring_spelled_by_purple_nodes(suffix_tree, node_colors)


def main():
    reader = sys.stdin
    text1 = next(reader).rstrip()
    text2 = next(reader).rstrip()
    result = find_longest_shared_substring(text1, text2)
    print(result)


if __name__ == '__main__':
    main()
